go/offline/rpc_cache.go

```diff
@@ -121,40 +126,96 @@ func (c *RPCCache) put(mctx libkb.MetaContext, version Version, rpcName string,
-	handler func(mctx libkb.MetaContext) (interface{}, error)) (err error) {
+// Serve an RPC out of the offline cache. The machinery only kicks
+// into gear if the `oa` OfflineAvailability mode is set to
+// BEST_EFFORT. If not, then just use the function `handler` which
+// does the main work of handling the RPC. Note that `handler` must
+// not modify anything in the caller's stack frame; it might be run in
+// a background goroutine after this function returns, to populate the
+// cache. `handler` also returns the return value for the RPC as an
+// interface, so it can be inserted into the offline cache in the
+// success case. We also pass this function a `version`, which will
+// tell the cache-access machinery to fail if the wrong version of the
+// data is cached. Next, we pass the `rpcName`, the argument, and the
+// pointer to which the result is stored if we hit the cache. This
+// flow is unfortunately complicated, but the issue is that we're
+// trying to maintain runtime type checking, and it's hard to do
+// without generics.
+//
+// If this function doesn't return an error, and the returned `res` is
+// nil, then `resPtr` will have been filled in already by the cache.
+// Otherwise, `res` should be used by the caller as the response.
+func (c *RPCCache) Serve(mctx libkb.MetaContext, oa keybase1.OfflineAvailability, version Version, rpcName string, encrypted bool, arg interface{}, resPtr interface{},
+	handler func(mctx libkb.MetaContext) (interface{}, error)) (res interface{}, err error) {
 
 	if oa != keybase1.OfflineAvailability_BEST_EFFORT {
-		_, err := handler(mctx)
-		return err
+		return handler(mctx)
 	}
 	mctx = mctx.WithLogTag("OFLN")
 	defer mctx.Trace(fmt.Sprintf("RPCCache#Serve(%d, %s, %v, %+v)", version, rpcName, encrypted, arg), func() error { return err })()
+
+	found, err := c.get(mctx, version, rpcName, encrypted, arg, resPtr)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we know we're not connected, use the cache value right away.
+	// TODO: API calls shouldn't necessarily depend on the
+	// connectivity as measured by gregor.
 	if mctx.G().ConnectivityMonitor.IsConnected(mctx.Ctx()) == libkb.ConnectivityMonitorNo {
-		found, err := c.get(mctx, version, rpcName, encrypted, arg, res)
-		if err != nil {
-			return err
-		}
 		if !found {
-			return libkb.OfflineError{}
+			return nil, libkb.OfflineError{}
 		}
-		return nil
+		return nil, nil // resPtr was filled in by get()
 	}
-	res, err = handler(mctx)
-	if err != nil {
-		return err
+
+	type handlerRes struct {
+		res interface{}
+		err error
 	}
-	tmp := c.put(mctx, version, rpcName, encrypted, arg, res)
-	if tmp != nil {
-		mctx.Warning("Error putting RPC to offline storage: %s", tmp.Error())
+	resCh := make(chan handlerRes, 1)
+
+	// New goroutine needs a new metacontext, in case the original
+	// gets canceled after the timeout below.
+	newMctx := libkb.NewMetaContextBackground(mctx.G())
```

Comment:
This'll kill the log tags.  Maybe try `mctx.BackgroundWithLogTags()`?

Reply:
or `mctx.BackgroundWithLogTags().WithCancel()` (see below)

Reply:
Oo, I was looking for something like `BackgroundWithLogTags` but somehow didn't find it.  Perfect, thanks!

Reply:
Not a big deal either way.